kratos/expression/integration_point_expression_io.cpp

```diff
@@ -0,0 +1,238 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Suneth Warnakulasuriya
+//
+
+// System incldues
+#include <type_traits>
+
+// Project includes
+#include "expression/expression_io_utils.h"
+#include "expression/variable_expression_data_io.h"
+#include "includes/process_info.h"
+
+// Include base h
+#include "integration_point_expression_io.h"
+
+namespace Kratos {
+
+namespace IntegrationPointExpressionIOUtils
+{
+
+template <class TContainerType, class TVariableType>
+static Expression::Pointer ReadToExpression(
+    TContainerType& rContainer,
+    TVariableType pVariable,
+    const DataCommunicator& rDataCommunicator,
+    const ProcessInfo& rProcessInfo)
+{
+    KRATOS_TRY
+
+    const IndexType number_of_entities = rContainer.size();
+
+    return std::visit([&rContainer, &rDataCommunicator, &rProcessInfo, number_of_entities](auto pVariable) {
+        using data_type = typename std::remove_const_t<std::remove_pointer_t<decltype(pVariable)>>::Type;
+        using raw_data_type = std::conditional_t<std::is_same_v<data_type, int>, int, double>;
+
+        // first get the shape correctly
+        std::vector<int> shape_info;
+        if (number_of_entities != 0) {
+            // initialize the shape with the first entity value
+            std::vector<data_type> output;
+            rContainer.begin()->CalculateOnIntegrationPoints(*pVariable, output, rProcessInfo);
+            VariableExpressionDataIO<std::vector<data_type>> variable_flatten_data_io(output);
+            const auto& shape = variable_flatten_data_io.GetItemShape();
+            shape_info.resize(shape.size() + 1, number_of_entities);
+            std::transform(shape.begin(), shape.end(), shape_info.begin() + 1, [](const auto v) -> int { return v; });
+        } else {
+            shape_info.resize(1, number_of_entities);
+        }
+
+        // now communicate the shape
+        const auto& r_shapes_info_in_ranks = rDataCommunicator.AllGatherv(shape_info);
+
+        // get the shape
+        std::vector<IndexType> shape;
+        for (const auto& r_shape_info_in_rank : r_shapes_info_in_ranks) {
+            if (r_shape_info_in_rank[0] != 0) {
+                shape.resize(r_shape_info_in_rank.size() - 1);
+                std::transform(r_shape_info_in_rank.begin() + 1, r_shape_info_in_rank.end(), shape.begin(), [](const auto v) -> IndexType { return v;});
+                break;
+            }
+        }
+
+        // remove number of entities for check
+        shape_info.erase(shape_info.begin());
```

Comment:
I'm completely lost. Please explain what the first entry in `shape_info` is supposed to be in a comment.

Reply:
done :)